### ISSUE TITLE
Breaking changes in ReWrite Plugin CoreDNS 1.8.4

### ISSUE THREAD

**Issue by miwithro** [coredns/coredns#4834]
AKS is currently running CoreDNS 1.8.3 and upgrading to CoreDNS 1.8.4 in September.  A breaking change with the rewrite plugin with CoreDNS 1.8.4 has been identified. 
 

This is an example for 1.8.3

```
<domain to be rewritten>.com:53 {
        errors
        cache 30
        rewrite name substring <domain to be rewritten>.com default.svc.cluster.local
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        forward .  /etc/resolv.conf # you can redirect this to a specific DNS server such as 10.0.0.10, but that server must be able to resolve the rewritten domain name
    }
```
The below is an example for 1.8.4 that won't break, which requires customers to update there configuration.
```
  <domain to be rewritten>.com:53 {
  log
  errors
  rewrite stop {
    name regex (.*)\.<domain to be rewritten>.com {1}.default.svc.cluster.local
    answer name (.*)\.default\.svc\.cluster\.local {1}.<domain to be rewritten>.com
  }
  forward . /etc/resolv.conf # you can redirect this to a specific DNS server such as 10.0.0.10, but that server must be able to resolve the rewritten domain name
```

**What happened**:  AKS is advising customers to update there rewrite plugin.

**What you expected to happen**: No breaking changes during patch version upgrades.

**How to reproduce it (as minimally and precisely as possible)**: Apply a configuration in 1.8.3 and upgrade to 1.8.4.

**Anything else we need to know?**:

**Environment**:
Azure Kubernetes Service 1.20+

- the version of CoreDNS: 1.8.4
- Corefile:
- logs, if applicable:
- OS (e.g: `cat /etc/os-release`):
- Others:

**Comment by chrisohaver** [comment]
The `answer` rewrite is required for both 1.8.3 and 1.8.4, otherwise a client may reject the answer due to the response name not matching the request name.

According to docs, the only time happens automatically is for `rewrite name exact` rules. https://github.com/coredns/coredns/blob/master/plugin/rewrite/README.md#response-rewrites

So for a `rewrite name substring` rule, request/response names will be mismatched unless a response rewrite option is appended.

**Comment by palma21** [comment]
Acknowledged that it was already required/documented even before, I guess the angst we're getting from customers here comes from the fact that "their config worked before and now it fails".

**Comment by chrisohaver** [comment]
Perhaps in 1.8.3, `rewrite name substring` rules errantly automatically adjusted the answer as `rewrite name exact` rules did. 
There _was_ some rewrite plugin code refactor in 1.8.4 that might have rectified the behavior to be in-line with the docs whilst being unaware of the prior undocumented behavior.  I'll dig into it to see if that's the case.

**Comment by chrisohaver** [comment]
OK - I verified that 1.8.3 does _not_ automatically rewrite answers for `rewrite name substring`.  The behavior appears to be consistent for both 1.8.3 and 1.8.4.

So something else must be in the mix.  

Re-reviewing the example Corefiles in the description, something seems fishy.  
1. The first example contains a kubernetes plugin, and the second does not. 
2. `rewrite name substring <domain to be rewritten>.com default.svc.cluster.local` doesn't seem like it accomplishes anything useful... `default.svc.cluster.local`  by itself is a namespace, which would be empty. It's not a complete name for something that has an IP address (e.g. lacking a service name  as in`kubernetes.default.svc.cluster.local`).
3. The 1.8.4 example does something entirely different...  Using `bar` as the domain, it converts things like `foo.bar.com` to `foo.default.svc.cluster.local`

Need more information here as to "what broke".

**Comment by robbiezhang** [comment]
@chrisohaver , here is a repo:

    reddog.microsoft.com:53 {
        log
        errors
        rewrite name substring reddog.microsoft.com google.com
        forward . /etc/resolv.conf
    }

I got the following error for 1.8.4    
nslookup www.reddog.microsoft.com
;; Question section mismatch: got www.google.com/A/IN
;; Question section mismatch: got www.google.com/A/IN
;; Question section mismatch: got www.google.com/A/IN
;; connection timed out; no servers could be reached

command terminated with exit code 1

But the same config works for 1.8.3

**Comment by chrisohaver** [comment]
> Need more information here as to "what broke".

Ah - I see I misunderstood how the `substring` option worked. 

I now understand "what broke".

Not sure _why_ it did yet...

**Comment by chrisohaver** [comment]
For reasons I cannot explain yet,  a client (I'm using `dig`) does _not_ error with `;; ;; Question section mismatch:` when querying a 1.8.3 CoreDNS that replies with mismatch question section.  But it does for 1.8.4.

```
; <<>> DiG 9.16.16 <<>> @127.0.0.1 -p 5390 foo.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 21337
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;foo.com.			IN	A

;; ANSWER SECTION:
example.com.		1366	IN	A	93.184.216.34

;; Query time: 32 msec
;; SERVER: 127.0.0.1#5390(127.0.0.1)
;; WHEN: Fri Aug 27 18:53:09 EDT 2021
;; MSG SIZE  rcvd: 63
```

But `foo.com` != `example.com`, so there _is_ a mismatch... 

I guess there must be something different in the response from 1.8.3 that somehow causes the client to not recognize the name mismatch.

**Comment by robbiezhang** [comment]
here is the tcpdump collected from the client. 
1.8.3: returns both the input (www.reddog.microsoft.com) and rewritten (www.google.com) DNS in the response.
![image](https://user-images.githubusercontent.com/9003737/131199884-379e663a-fef0-45d4-8f6f-70182e8c47f5.png)

1.8.4: returns rewritten DNS (www.google.com) twice in the response.
![image](https://user-images.githubusercontent.com/9003737/131200003-827c9981-5945-485c-8404-36d61ec2c469.png)

Sample config:

```
  reddog.microsoft.com:53 {
    log
    errors
    rewrite name substring reddog.microsoft.com google.com
    forward . /etc/resolv.conf
  }
```

**Comment by ykhr53** [comment]
Found a huge commit on rewrite plugin, please refer https://github.com/coredns/coredns/commit/40edf1e566a0b3709139420aeba542d9b788b1d0.
I tested reverting the commit on my dev env (CoreDNS 1.8.4), and I confirmed it gets back to working fine after the revert.
So it seems the commit https://github.com/coredns/coredns/commit/40edf1e566a0b3709139420aeba542d9b788b1d0 caused the behavior change, although I haven't read through it.

**Comment by chrisohaver** [comment]
> Found a huge commit on rewrite plugin, please refer [40edf1e](https://github.com/coredns/coredns/commit/40edf1e566a0b3709139420aeba542d9b788b1d0).
> I tested reverting the commit on my dev env (CoreDNS 1.8.4), and I confirmed it gets back to working fine after the revert.
> So it seems the commit [40edf1e](https://github.com/coredns/coredns/commit/40edf1e566a0b3709139420aeba542d9b788b1d0) caused the behavior change, although I haven't read through it.

Yes that’s the refactor I referred to above. Based on the tcpdump it looks like it unintentionally changed how it handles the question section in the response, no longer retaining the original pre-rewrite question.

Needs to be fixed and unit tests added to catch this. But I won’t have time to look at this in more depth until Monday.

**Comment by chrisohaver** [comment]
Cc @mandelsoft

**Comment by chrisohaver** [comment]
Proposed fix: #4840

**Comment by chrisohaver** [comment]
>The below is an example for 1.8.4 that won't break, which requires customers to update there configuration.
>
>```
>  <domain to be rewritten>.com:53 {
>  log
>  errors
>  rewrite stop {
>    name regex (.*)\.<domain to be rewritten>.com {1}.default.svc.cluster.local
>    answer name (.*)\.default\.svc\.cluster\.local {1}.<domain to be rewritten>.com
>  }
>  forward . /etc/resolv.conf # you can redirect this to a specific DNS server such as 10.0.0.10, but that server must be able to resolve the rewritten domain name
>```

FWIW, For 1.8.4, a much simpler workaround is to the use the `answer auto` option (new in 1.8.4) ... e.g. 

`rewrite name substring <domain to be rewritten>.com default.svc.cluster.local answer auto`

**Comment by robbiezhang** [comment]
Chris, in AKS, the coredns version is controlled by AKS, while the config is partially controlled by customer. We find this new setting `answer auto` for 1.8.4 which would fix the issue. However, as you mentioned, it's not supported in 1.8.3, and causes the coredns crash. 
Therefore, we provide the `answer name` solution which works in both 1.8.3 and 1.8.4, and notify the customer to upgrade their config, before we upgrade the coredns to 1.8.4.
We hope new coredns is backward compatible to the old config, and the default behavior should be the same as the old one.

**Comment by chrisohaver** [comment]
> We hope new coredns is backward compatible to the old config, and the default behavior should be the same as the old one.

Yes, it should.  Note however, that with the prior configuration _some_ client resolvers will complain about the name mismatch, even in 1.8.3.  

More detail ... There are 3 types of the "question name" at play.

- The original question that the client made - taken from the Question section of the request message.
- **Response Question Name**: From the Question section of the response message. 
- **Answer Record Name**: From the names of records in the Answer section of the response message

Most client resolvers will fail when the original question != **Response Question Name**.  Hence this was always restored in the response so they would be equal (until 1.8.4 by accident).

After early versions of the _rewrite_ plugin had been released for some time, we got reports that _some_ client resolvers were failing when the original question != **Answer Record Name**, so the `answer` rewrite option was added (long ago, in 1.0.5).
